### PR TITLE
fix: send a dlog proof for h1 wrt h2

### PR DESCRIPTION
**Pull request by milapsheth**
- Prove that the order of `h1` and `h2` are equal for the Zksetup. The previous proof only proved that `h2` is in the subgroup `<h1>`. This bug was reported in this [blog post](https://medium.com/velasblockchain/counter-strike-threshold-attack-87f3b456b1e0) for another implementation.
- Update the max message sizes to accommodate the new proof.

### CODE REVIEW

File: src/gg20/crypto_tools/paillier/zk/composite_dlog.rs
Comment (ggutoski) [inline src/gg20/crypto_tools/paillier/zk/composite_dlog.rs:152]: Why only a debug assert for this?  I guess if we don't want to return an error then we can't guard against invalid witness.

File: src/gg20/crypto_tools/paillier/zk/composite_dlog.rs
Comment (ggutoski) [reply src/gg20/crypto_tools/paillier/zk/composite_dlog.rs:201]: It's not a problem though, right?  ie. if someone sends a `r` that's longer than it needs to be then the proof is still valid and will be accepted, right?  In that case I don't see a problem---any reasonable sanity check is sufficient.

File: src/gg20/crypto_tools/paillier/zk/mod.rs
Comment (ggutoski) [inline src/gg20/crypto_tools/paillier/zk/mod.rs:86]: This is a usability gotcha.  In order to use `witness2` you need to know to compose your own `stmt2` from the `stmt1` you have by swapping `g`, `v`.  At the very least we need to document this.  But perhaps a better solution is to have `CompositeDLogStmt::setup` just return `(stmt1, witness1, stmt2, witness2)`.

This return tuple is getting large though.  Perhaps it's time to bundle these statements/witnesses into a larger struct?

**Review by ggutoski — COMMENTED** [review]
A few minor comments.  I have a couple commits of my own that I'll push immediately.

File: src/gg20/crypto_tools/paillier/zk/composite_dlog.rs
Comment (milapsheth) [inline src/gg20/crypto_tools/paillier/zk/composite_dlog.rs:126]: I realized my comment was incorrect. Since we're computing `s^(-1) mod phi(N)`, you want `s` to be invertible `mod phi(N)` and not `mod N`. I'll update it.

File: src/gg20/crypto_tools/paillier/zk/composite_dlog.rs
Comment (milapsheth) [inline src/gg20/crypto_tools/paillier/zk/composite_dlog.rs:62]: Thanks, I tried using `max` but it was complaining about not being a `const` function. It seems I was using the `std::cmp::max` function which is different than the one already in scope.

File: src/gg20/crypto_tools/paillier/zk/composite_dlog.rs
Comment (milapsheth) [reply src/gg20/crypto_tools/paillier/zk/composite_dlog.rs:169]: All our zkproofs currently assume that the statement and witness is already valid (which is the case currently). I have added similar `debug_asserts` only as a sanity check for testing. I would prefer to avoid calling more expensive operations like `modpow` as a sanity check in release mode.

File: src/gg20/crypto_tools/paillier/zk/composite_dlog.rs
Comment (milapsheth) [inline src/gg20/crypto_tools/paillier/zk/composite_dlog.rs:242]: Ah, that makes more sense. Neat.

File: src/gg20/crypto_tools/paillier/zk/mod.rs
Comment (milapsheth) [reply src/gg20/crypto_tools/paillier/zk/mod.rs:86]: Yeah, I'll create a method in `CompositeDLogStmt` instead, since we need to do this in `verify` too, where `::setup` isn't being called.
Alternatively, we could modify the zk proof to prove two things internally, so the proof itself will have `y1` and `y2` elements (this would also save some time by not rechecking everything again). The downside is that the zk proof is not as standalone and doesn't simply mimic the paper.

File: src/gg20/crypto_tools/paillier/zk/composite_dlog.rs
Comment (milapsheth) [inline src/gg20/crypto_tools/paillier/zk/composite_dlog.rs:150]: I think if `s` is invertible `mod phi(N)` and if `g` is asymmetric (quadratic nonresidue), then `v` should always be asymmetric too: if there was `x` such that `x^2 = v = g^(-s) mod N`, then `g = (x^(-s^(-1)))^2 mod N` making `g` a quadratic residue. But the paranoid part of me left this in.

Comment (milapsheth) [reply src/gg20/crypto_tools/paillier/zk/composite_dlog.rs:150]: So, to be asymmetric `v` would have to be quadratic nonresidue mod only one of `p` or `q`. making it a nonresidue mod `N`. It could be that it's a quadratic nonresidue mod both `p` and `q` perhaps?

File: src/gg20/crypto_tools/paillier/zk/composite_dlog.rs
Comment (milapsheth) [inline src/gg20/crypto_tools/paillier/zk/composite_dlog.rs:88]: This is to `zeroize` on `drop`.

File: src/gg20/crypto_tools/paillier/zk/composite_dlog.rs
Comment (ggutoski) [reply src/gg20/crypto_tools/paillier/zk/composite_dlog.rs]: Well, we could add a `MAX_R_MASK_SIZE` const generic parameter to `CompositeDLogStmt`.  Then the two statements/proofs would be of different types and each would use the right constant.

File: src/gg20/crypto_tools/paillier/zk/composite_dlog.rs
Comment (ggutoski) [reply src/gg20/crypto_tools/paillier/zk/composite_dlog.rs:88]: I'm not a huge fan of this.  It's not clear why you're using `Randomness` here.  Perhaps your github PR comment should instead be a code comment. 😛 A proper solution would make the purpose clear.  (Sigh)  Do we need something like `Randomness` except it's called `ZeroizeOnDrop`?

**Review by ggutoski — APPROVED** [review]
There are 1 or 2 dangling comments but I might not be available before we need to merge, so I'll just approve and leave it to you to merge at your discretion.  Worst case you can merge as-is and we can open new issues for any comments that demand more work.

File: src/gg20/crypto_tools/paillier/zk/mod.rs
Comment (ggutoski) [reply src/gg20/crypto_tools/paillier/zk/mod.rs]: I like the new `get_inverse_statement` method.

File: src/gg20/crypto_tools/paillier/zk/composite_dlog.rs
Comment (milapsheth) [reply src/gg20/crypto_tools/paillier/zk/composite_dlog.rs]: I thought that might interfere with `serde`, but it makes sense now. I have added that.

**Comment by milapsheth** [comment]
> There are 1 or 2 dangling comments but I might not be available before we need to merge, so I'll just approve and leave it to you to merge at your discretion. Worst case you can merge as-is and we can open new issues for any comments that demand more work.

Sure, I'll merge this, and let's discuss if we can improve the design more next week.

File: src/gg20/crypto_tools/paillier/zk/composite_dlog.rs
Comment (ggutoski) [reply src/gg20/crypto_tools/paillier/zk/composite_dlog.rs]: Could use the Secret crate from rustcrypto

**Pull request merged**
